GUI to Edit EAV Attributes & Sets - Base Code + Category

justinbeaty opened the pull request:
<!---
    Thank you for contributing to OpenMage LTS.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
This PR gives backend users the possibility to create/edit/delete any EAV attribute. In other words, all of these:

```sql
mysql> SELECT entity_type_id,entity_type_code FROM eav_entity_type ORDER BY entity_type_id;
+----------------+------------------+
| entity_type_id | entity_type_code |
+----------------+------------------+
|              1 | customer         |
|              2 | customer_address |
|              3 | catalog_category |
|              4 | catalog_product  |
|              5 | order            |
|              6 | invoice          |
|              7 | creditmemo       |
|              8 | shipment         |
+----------------+------------------+
```

I have only done light testing on this, we probably need more work before this can be merged as a useful feature.

### Related Pull Requests
<!-- related pull request placeholder -->
This PR builds on top of the work done by @fballiano. It is meant to supersede these PRs if the community accepts.
1. #2260
2. #2264

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format OpenMage/magento-lts#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Open the backend
2. See new options in Sales > Attributes, Catalog > Attributes, Customer > Attributes
3. Try to create new attributes or attribute sets

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->
Todo:
- [x] Check translation strings
- **For another PR**: Modify customer attributes to add `use_in_forms` option?
- **For another PR**: Modify Customers > Manage Groups to allow us to select a attribute set for each customer group?
- **For another PR**: Block the user from editing/deleting "system" attributes. Use `adminhtml_[entityCode]_attribute_scopes` event to disallow changing scope perhaps?
- [x] Check ACL permissions, and make sure they didn't break BC with new Catalog > Attributes > Products > Attributes path.
- [x] Fit icon-head for `admin/XXX_attribute/new/` and `admin/XXX_set/add/`
- [ ] Figure out Copyright headers -- most of the code is copy/pasted from other core code, so I think we still need Magento, LLC copyright notice (in addition to ours in another PR).
- [x] Fix array short syntax
- [ ] Remove all prototypejs code
- [x] merge at least customer or category management PR into this, otherwise this is untestable
- **For another PR**: Review the SQL commands at the end of [this comment](https://github.com/OpenMage/magento-lts/pull/2264#issuecomment-1181890693). Should we add an upgrade script?


### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All automated tests passed successfully (all builds are green)
 - [x] Add yourself to contributors list
 <!--- 
    Install: `yarn add --dev all-contributors-cli`
    Add yourself: `yarn all-contributors add @YOUR_NAME <types>`
    This updates `.all-contributorsrc, README.md` and commits this changes automatically
    contribution types: code, doc, design
    See other contributions type at https://allcontributors.org/docs/en/emoji-key
 -->

justinbeaty commented:
The basic code structure is that we have generic blocks in Mage_Eav that make grids and edit forms for any EAV attribute. To determine which one we are editing, we read from the registry. 

Then we also have an abstract controllers that take care of the grid/edit/save/etc actions. For each EAV attribute we want to allow the user to edit, we just add the menu item and extend from the abstract controllers.

justinbeaty commented:
Also for discussion, what is the reason someone would edit any of the Sales (Orders, Invoice, Creditmemo, Shipment) EAV attributes, or attribute sets? Has anyone ever done anything with these EAV entities in custom code? Would we want to show extra attributes in admin create order screens?

fballiano commented:
> Also for discussion, what is the reason someone would edit any of the Sales (Orders, Invoice, Creditmemo, Shipment) EAV attributes, or attribute sets? Has anyone ever done anything with these EAV entities in custom code? Would we want to show extra attributes in admin create order screens?

I did that a lot of times in the past 12 years :-D
but only attributes and not sets, for those entities

fballiano commented:
I love this PR but I think there's a problem, orders/invoice etc have flat tables with static attributes (meaning columns in the flat table), not in the EAV. If I add an order attribute I want to have it on the flat table (maybe even on the _grid table) but not in the EAV...

I also don't know how much sense would the attribute_set feature have, since they're _flat tables... 

Same thing for the customer attribute sets, I'm not sure to force user to choose an attribute set for every customer group.. I'd leave that part for a future implementation (and that's why I didn't include it in my original PRs)

justinbeaty commented:
> > Also for discussion, what is the reason someone would edit any of the Sales (Orders, Invoice, Creditmemo, Shipment) EAV attributes, or attribute sets? Has anyone ever done anything with these EAV entities in custom code? Would we want to show extra attributes in admin create order screens?
> 
> I did that a lot of times in the past 12 years :-D but only attributes and not sets, for those entities
>
> I love this PR but I think there's a problem, orders/invoice etc have flat tables with static attributes (meaning columns in the flat table), not in the EAV. If I add an order attribute I want to have it on the flat table (maybe even on the _grid table) but not in the EAV...
> 
> I also don't know how much sense would the attribute_set feature have, since they're _flat tables... same thing for the "scope".

I add columns to the flat table quite often as well, I just wasn't thinking of it as using EAV attributes, but I guess it is technically? Did Magento used to use the EAV table for orders before they decided on the flat tables for performance?

Anyway, so should we just remove all Sales > Attributes menu for now? It's as easy as dropping that one commit.

-----

> Same thing for the customer attribute sets, I'm not sure to force user to choose an attribute set for every customer group.. I'd leave that part for a future implementation (and that's why I didn't include it in my original PRs)

I was thinking similar @kiatng's comment in the other PR:

> I have a similar custom extension, including the _Manage Attribute Sets_ for organizing the attributes for different customer groups.

Where customer groups would by default use the default attribute set, so the user would only need to change it if they wanted to. But more than that, the custom attributes do not even show up anywhere in the Customer > Edit screen yet, we need the `use_in_forms` field controllable.

kiatng commented:
All sales entities do not use EAV. There is no trace of it except in the table `eav_entity_type`:
![image](https://user-images.githubusercontent.com/1106470/178862213-e8497d86-e59f-43a6-85a0-c25c3439b4a4.png)

I think the last 4 rows were forgotten and are now ghosts. The collections of the sales entities do not extend the EAV class, it has its own abstract class for BC to EAV:

https://github.com/OpenMage/magento-lts/blob/042d4e5c979776debf97843e80b5c5aaf8b380e4/app/code/core/Mage/Sales/Model/Resource/Collection/Abstract.php#L57-L68

So, only the catalog and customer module use EAV. Recently, I have implemented a custom EAV module. So, for me, I can see a need to abstract common functions to the EAV module instead of extending to the catalog module, and as pointed out by @fballiano, create unnecessary dependency. 

I use attribute sets to organize the attributes for different customer groups. The project has over 200 customer attributes, each customer group having very different set of attributes. 

My concern is that changes to EAV could make it very difficult to merge OM to existing projects with custom EAV.

justinbeaty commented:
So let's definitely just remove the catalog EAV grids. Technically you could add attributes & sets for orders, invoices, etc, but with no way to retrieve those values in the collections it's pointless and confusing.

> I use attribute sets to organize the attributes for different customer groups. The project has over 200 customer attributes, each customer group having very different set of attributes.
> 
> My concern is that changes to EAV could make it very difficult to merge OM to existing projects with custom EAV.

The PR as it now should definitely not conflict with anything since it's all new code, but I see your point if we hook up a customer group -> customer attribute set relation. Still, I think it could be worth it maybe just waiting to merge this PR until the next major version.

How did you correlate attribute sets with customer groups? My first thought is to modify the `customer_group` table:
```
mysql> describe customer_group;
+---------------------+-------------------+------+-----+---------+----------------+
| Field               | Type              | Null | Key | Default | Extra          |
+---------------------+-------------------+------+-----+---------+----------------+
| customer_group_id   | smallint unsigned | NO   | PRI | NULL    | auto_increment |
| customer_group_code | varchar(32)       | NO   |     | NULL    |                |
| tax_class_id        | int unsigned      | NO   |     | 0       |                |
+---------------------+-------------------+------+-----+---------+----------------+
3 rows in set (0.00 sec)
```

To add something like `customer_attribute_set_id` and maybe even `customer_address_attribute_set_id`. Probably both are useful in the case you want to control both attribute sets per group.

If we can do it in a way that is maybe compatible with what you've done, then that's at least one site that wouldn't break.

-----

Okay so next question, is about attribute sets on categories. I can't see a need for this, I just included it in the PR since it was only a few lines to do so. How should the menu entries look like? A few options:

* Catalog
  * Manage Products
  * Manage Categories
  * Attributes
    * Product 
      * Attributes
      * Attribute Sets
    * Category
      * Attributes

* Catalog
  * Manage Products
  * Manage Categories
  * Attributes
    * Product Attributes
    * Product Attribute Sets
    * Category Attributes

* Catalog
  * Manage Products
    * Attributes
    * Attribute Sets
  * Manage Categories
    * Attributes
  * ~~Attributes~~

Option 1 is like I have it now in this PR, just with the Category Attribute Sets option removed. Option 2 is like @fballiano had it but with the word "Product" added to the first two options. Option 3 is a bit weird because I don't think there's other cases where a clickable menu item has children.

justinbeaty commented:
> So if the only EAV entities are actually customer, customer_address, catalog and category maybe my approach wasn't so bad after all :-D
> 
> but... somebody could have custom EAV entities and it would be awesome to manage their attributes with a GUI anyway...
> 
> so maybe we could remove the 4 non-EAV entities from this PR and postpone it after merging this one?

Yeah, your original PRs were definitely on the money here. I just thought it could be abstracted and then I added all the EAV entities since it only requires adding the controller & adminhtml.xml entries.

I definitely would consider now to make my own EAV entities though, since the GUI boilerplate is taken care of.

I reverted the catalog grids in a commit, I am not sure what you mean by the second part (postpone what?)

P.S. I will squash and force push some of these commits before this PR is merged (if it is merged). I just didn't want to keep force-pushing in case anyone has cloned this branch.

fballiano commented:
> Okay so next question, is about attribute sets on categories. I can't see a need for this, I just included it in the PR since it was only a few lines to do so. How should the menu entries look like? A few options:

I category_attribute_set should be tied to the store, since you may want to have different attribute sets for categories in different websites (but I'd use store, not store_views, since the store is the root category).

> Option 1 is like I have it now in this PR, just with the Category Attribute Sets option removed. Option 2 is like @fballiano had it but with the word "Product" added to the first two options. Option 3 is a bit weird because I don't think there's other cases where a clickable menu item has children.

I wouldn't change the products attribute admin path that everybody has now in production, since we'd need an update script to fix the ACL. So I think I'd go for option 2.

fballiano commented:
> I reverted the catalog grids in a commit, I am not sure what you mean by the second part (postpone what?)

I mean that this PR shouldn't do anything about order/quote/invoice etc

justinbeaty commented:
> > Okay so next question, is about attribute sets on categories. I can't see a need for this, I just included it in the PR since it was only a few lines to do so. How should the menu entries look like? A few options:
> 
> I category_attribute_set should be tied to the store, since you may want to have different attribute sets for categories in different websites (but I'd use store, not store_views, since the store is the root category).

+1, that makes sense to me. Is it too much to add in to this PR the code to associate category attribute sets to stores, and customer/customer address attribute sets to customer groups? Should that be a follow up PR?

> > Option 1 is like I have it now in this PR, just with the Category Attribute Sets option removed. Option 2 is like @fballiano had it but with the word "Product" added to the first two options. Option 3 is a bit weird because I don't think there's other cases where a clickable menu item has children.
> 
> I wouldn't change the products attribute admin path that everybody has now in production, since we'd need an update script to fix the ACL. So I think I'd go for option 2.

If we allow the Category Attribute Sets option, then I am leaning towards this ideally:

* Catalog
  * Manage Products
  * Manage Categories
  * Attributes
    * Product
      * Attributes
      * Attribute Sets
    * Category
      * Attributes
      * Attribute Sets

To fix the BC break, I made PR #2321 that should allow us to change the ACL & menu structure without changing the ACL path.

fballiano commented:
> +1, that makes sense to me. Is it too much to add in to this PR the code to associate category attribute sets to stores, and customer/customer address attribute sets to customer groups? Should that be a follow up PR?

I'd do it in a separate PR cause I think it's not everybody's use case and one may want to add/edit attributes and write custom code to handle them on the frontend side
 
> * Catalog
>   
>   * Manage Products
>   * Manage Categories
>   * Attributes
>     
>     * Product
>       
>       * Attributes
>       * Attribute Sets
>     * Category
>       
>       * Attributes
>       * Attribute Sets

To me it's not super necessary to change the path and introduce that other PR to handle that.
I'd simpy do
- catalog
- - manage products
- - manage categories
- - attributes
- - - manage product attributes
- - - manage product attribute sets
- - - manage category attributes
- - - manage cateogyr attribute sets

it's never extremely comfortable to navigate many sublevels plus there would be only 4 items in the "attributes" one, not a very messy situation ;-)

justinbeaty commented:
I pushed the menu changes. A few other TODOs:

## Block the user from editing/deleting "system" attributes

Should we do this? It seems like we are not blocked from editing system product attributes. I suppose an admin could probably do some damage by editing those items, but I do not know what will happen.

## Review the SQL commands at the end of this comment: https://github.com/OpenMage/magento-lts/pull/2264#issuecomment-1181890693

```sql
UPDATE eav_attribute SET frontend_input = 'boolean' WHERE source_model = 'eav/entity_attribute_source_boolean' AND entity_type_id = 3;

DELETE eag.* FROM eav_attribute_set eas LEFT JOIN eav_attribute_group eag ON eas.attribute_set_id = eag.attribute_set_id WHERE eas.entity_type_id = 3 AND eag.attribute_group_name = 'General';
UPDATE eav_attribute_set eas LEFT JOIN eav_attribute_group eag ON eas.attribute_set_id = eag.attribute_set_id SET eag.default_id = 1 WHERE eas.entity_type_id = 3 AND eag.attribute_group_name = 'General Information';
```

I think it's a bug that we had `frontend_input = 'boolean'` but not `source_model = 'eav/entity_attribute_source_boolean'`. So if you went to edit some of the system attributes (which you should probably not do) then you could add options to a yesno type. That obviously makes no sense.

Also, we can clean up some the unused attribute group name "General" for the default product category attribute set.

---

I appreciate the feedback and of course your original initiative on this @fballiano, I do not want to let this PR stagnate so I'm trying to work through all the issues. :)

fballiano commented:
it's possible to edit system attributes but not delete them, I think we should have the same policy here.

Although... let's take "media_gallery" attribute as an example... I can't imagine what's happening if you modify the "scope" of this attribute but it's been like this since the beginning of magento so...

fballiano commented:
About the SQL query review, I couldn't test it but maybe it should be a PR on its own?

justinbeaty commented:
@fballiano It shouldn't be possible right now to delete system attributes because of this line:

https://github.com/OpenMage/magento-lts/blob/74347684d15c0cf376394874fc06dfefcbc06308/app/code/core/Mage/Eav/Block/Adminhtml/Attribute/Edit.php#L51-L55

I don't know what would happen either if someone totally messes up system attributes by changing the scope or other properties of the attribute. But since it's possible to edit system attributes for the catalog_product entity, I think and changes preventing that should be in another PR.

I'll knock a few other TODOs from my list, squash & force push this branch to clean it up some.

fballiano commented:
I tested this morning, although not super extensively, but love this!

fballiano commented:
@justinbeaty do you think we should target v20 instead of v19? just to try to drive more attention to v20?

justinbeaty commented:
Updated the base branch to 20.0. Running through some tests now, I'll have some changes to push.

justinbeaty commented:
Updated the base branch to 20.0 and pushed [some changes](https://github.com/OpenMage/magento-lts/compare/dbaead2e4533c79864a08b0c13c634e74ff4573c..6afbab4a1d2223963044da3c4008c7a1512616af).

I tested adding one of each attribute frontend type to a category set, and tried to load the edit category page. The `price` field caused an error which I suspected it would; I had fixed this same error for the `boolean` type already.

In any case, I decided to remove `price` and `media_image` from the base EAV code. There's an event fired to be able to modify this for specific attributes. I also moved the boolean rendered from `Adminhtml/Block/Catalog/Category/Helper/Boolean.php` to `Eav/Block/Helper/Form/Boolean.php` to make it more generic.

I also documented these new events in EVENTS.md. I put 22.0.0 as the version, not sure what it will actually be merged into so I'll change that before the final merge.

| Event name | With Magento or since OpenMage |
| ---------- | -------------- |
| adminhtml_[entityCode]_attribute_types | 22.0.0 |
| adminhtml_[entityCode]_attribute_scopes | 22.0.0 |
| adminhtml_[entityCode]_attribute_edit_prepare_form | 22.0.0 |

Hopefully that about fixes everything needed in this PR.

fballiano commented:
beautiful job :-) and I hope it will be merged much sooner than v22 ;-)

justinbeaty commented:
@fballiano I didn't mean that it should wait that long, I just wasn't sure if the next major release is v21 or v22.

justinbeaty commented:
> > Okay so next question, is about attribute sets on categories. I can't see a need for this, I just included it in the PR since it was only a few lines to do so. How should the menu entries look like? A few options:
> 
> I category_attribute_set should be tied to the store, since you may want to have different attribute sets for categories in different websites (but I'd use store, not store_views, since the store is the root category).
> 
> >

I was playing around with this a bit, but I'm not sure if this works because two stores could share the same root category. Maybe instead you can select the attribute set when clicking "Add Root Category" in the Manage Categories screen.

fballiano commented:
> I was playing around with this a bit, but I'm not sure if this works because two stores could share the same root category. Maybe instead you can select the attribute set when clicking "Add Root Category" in the Manage Categories screen.

makes sense, the "store" was my first idea since that's how I've seen it using similar things (in multi website environments) but to tie it to the root category is maybe better (although maybe makes it a bit more difficult to know which attribute set is used by a child category?)

justinbeaty commented:
A user on Discord mentioned this module: https://github.com/r-martins/Clarion_Customeratributes. It looks like it has support for assigning attributes to forms. I'm just noting it here so I can try installing it as well as this PR and see if there's any conflicts or issues.

They also mentioned there's a similar one for category attributes, but I haven't found it yet.

fballiano commented:
I think I have code for assigning customer attributes to forms if we want to add it already (since we don't have others reviews and we can't merge yet).

justinbeaty commented:
@fballiano I've done it this way in other modules:

```php
$attribute = Mage::getSingleton('eav/config')->getAttribute('customer', $attribute_code);
$attribute->setData('used_in_forms', array('adminhtml_customer'));
$attribute->save();
```

And it looks like these are the various options:

```
mysql> select distinct(form_code) from customer_form_attribute;
+----------------------------+
| form_code                  |
+----------------------------+
| adminhtml_checkout         |
| adminhtml_customer         |
| adminhtml_customer_address |
| checkout_register          |
| customer_account_create    |
| customer_account_edit      |
| customer_address_edit      |
| customer_register_address  |
+----------------------------+
8 rows in set (0.01 sec)
```

We could certainly add that functionality into this PR, otherwise I figured it might go into a separate PR as marked in the TODO in the first comment:

> * For another PR: Modify customer attributes to add use_in_forms option?
> * For another PR: Modify Customers > Manage Groups to allow us to select a attribute set for each customer group?

In either case, I think we can easily do it via an observer to modify the edit form and add a multiselect field.

justinbeaty commented:
If several PRs is a better idea, which I sort of think it is to not make this PR too big or take too long to be merged, we can also remove the following two commits from this PR and make them the base of separate PRs each:

[Mage_Customer: add EAV grids](https://github.com/OpenMage/magento-lts/pull/2317/commits/a4bbceb287e549af0ee420ce230deafca1d96f0f) 

[Mage_Catalog: add EAV grids](https://github.com/OpenMage/magento-lts/pull/2317/commits/f03fa3c9138c505fd59e53b3bfef65d780c17459)

So this PR would be just adding the EAV grid code (but not using them anywhere) and we can have two separate PRs (one for customer and one for categories) where we can discuss how they interact with customer groups, root categories, etc.

fballiano commented:
that's fine by me (I actually proposed that same idea in my former PRs), I was saying that we could continue to add feature since there's no other review. I'll see if I can do something about that.

fballiano commented on EVENTS.md:
```diff
@@ -36,6 +36,9 @@
 | admin_user_validate | 1.9.4.5 |
+| adminhtml_[entityCode]_attribute_types | 22.0.0 |
+| adminhtml_[entityCode]_attribute_scopes | 22.0.0 |
+| adminhtml_[entityCode]_attribute_edit_prepare_form | 22.0.0 |
```
I'd use the next v20 version number for this, if it won't me merged in time we'll re-change it later, otherwise we can't merge with 22.0.0 since it doesn't exist.

justinbeaty replied:
Updated to 20.0.15, although I think flyingmana said he's going to do a release this week so it might need to be updated again.

justinbeaty commented:
@fballiano I removed both the controllers and the menu config for the customer and catalog grids. My thoughts is it allows this PR to be merged sooner while we discuss specifics about the customer / catalog implementations. I think this PR is by itself a standalone feature since it allows a programmers to implement their own EAV grid.

Along the same lines, I thought it wouldn't be great to have this PR as it was before (with bare bones customer / category grids) and then have to make another PR to implement things like `use_in_forms` or the customer group / attribute set integration. This way, once either the category / customer code is fully developed it can be merged as one polished feature.

Does that make sense? That is what I meant to say in [this comment](https://github.com/OpenMage/magento-lts/pull/2317#issuecomment-1198591221) but maybe I wasn't very clear.

fballiano commented:
Me personally I wouldn't suggest to people to write their own grids etc, M1 were based on EAV for extendability but it the GUI wasn't provided because it was only in the enterprise edition. But OM should have that feature (again, only in my opinion) and it should be as complete as possible.

But, as it was before, a user could easily create attributes and then modify the template (which all of us are pretty used to).

When I used M1 enterprise we had to modify the used_in_forms anyway since sometimes the way it was implemented it wouldn't totally adapt to the website layout so.. in my opinion... it is not that bad if left out ;-)

I wouldn't know how to test this PR, I'd have to write a script which... more difficult for sure.

Again, just my opinions.

justinbeaty commented:
A couple of thoughts:

> Me personally I wouldn't suggest to people to write their own grids etc,

People definitely should not have to write the customer or category grids, we will supply those. I meant if they decide to create their own EAV entity, then they could easily spin up the GUI with this PR.

> But OM should have that feature (again, only in my opinion) and it should be as complete as possible. [...] When I used M1 enterprise we had to modify the used_in_forms anyway since sometimes the way it was implemented it wouldn't totally adapt to the website layout so.. in my opinion... it is not that bad if left out ;-)

I feel this is conflicting, since if we want it to be as complete as possible, we should include use_in_forms capability as well as some sort of functionality to make the attribute sets useful (else why include the GUI to edit them?) 

My thought was that if this PR was merged before I split it into three separate PRs, then users might be confused about some features that seem useless, such as creating new attribute sets on either customer or categories.

> I wouldn't know how to test this PR, I'd have to write a script which... more difficult for sure.

It is still the same first commit as was before in this PR. The other two PRs also include this commit so it could be tested that way.

---

I'm for whatever makes it easier for the maintainers and whatever sticks closest to your original vision / PRs that you opened. I thought you had agreed on the splitting of the PRs but I may have misunderstood.

fballiano commented:
> I feel this is conflicting, since if we want it to be as complete as possible, we should include use_in_forms capability as well as some sort of functionality to make the attribute sets useful (else why include the GUI to edit them?)

I was meaning that creating attributes programmatically is not the easiest thing, add them in a template is much easier. Anyway obviously we should have the use_in_forms, although to me it's not necessary in the first second, while to me all the rest it is.

> It is still the same first commit as was before in this PR. The other two PRs also include this commit so it could be tested that way.

I know... I wasn't talking about me me ;-)

> I'm for whatever makes it easier for the maintainers and whatever sticks closest to your original vision / PRs that you opened. I thought you had agreed on the splitting of the PRs but I may have misunderstood.

I'm super sorry I don't know what I understood 😅 but anyway please don't think like decide stuff here in OM, I'm just a contributor expressing his opinions 😊 so you should always do what you feel is right.

justinbeaty commented:
@fballiano well let me see what I can accomplish with use_in_forms over the next day or so. I also didn't want to keep pushing commits to this branch in case other maintainers are trying to review.

And not problem at all on the misunderstanding, even if you're just a contributor too, it was your original idea so I value your opinion on this highly.

justinbeaty commented on app/code/core/Mage/Eav/controllers/Adminhtml/Attribute/AbstractController.php:
```diff
@@ -0,0 +1,346 @@
+                $data['backend_model'] = $helper->getAttributeBackendModelByInputType($data['frontend_input']);
+            }
+
+            if (!isset($data['is_configurable'])) {
+                $data['is_configurable'] = 0;
+            }
+            if (!isset($data['is_filterable'])) {
+                $data['is_filterable'] = 0;
+            }
+            if (!isset($data['is_filterable_in_search'])) {
+                $data['is_filterable_in_search'] = 0;
+            }
```
dead code

justinbeaty commented on app/code/core/Mage/Eav/controllers/Adminhtml/Attribute/AbstractController.php:
```diff
@@ -0,0 +1,346 @@
+                $data['default_value'] = $this->getRequest()->getParam($defaultValueField);
+            }
+
+            if(!isset($data['apply_to'])) {
+                $data['apply_to'] = array();
+            }
```
more dead code

justinbeaty commented on app/code/core/Mage/Eav/controllers/Adminhtml/Attribute/AbstractController.php:
```diff
@@ -0,0 +1,346 @@
+            try {
+                $model->save();
+                $session->addSuccess(
+                    Mage::helper('eav')->__('The product attribute has been saved.'));
```
should be "The attribute has been saved."

justinbeaty commented on app/code/core/Mage/Eav/controllers/Adminhtml/Attribute/AbstractController.php:
```diff
@@ -0,0 +1,346 @@
+            try {
+                $model->delete();
+                Mage::getSingleton('adminhtml/session')->addSuccess(
+                    Mage::helper('eav')->__('The product attribute has been deleted.'));
```
Should be "The attribute has been deleted."

justinbeaty commented on app/code/core/Mage/Eav/Block/Adminhtml/Attribute/Edit/Tab/Main.php:
```diff
@@ -0,0 +1,102 @@
+class Mage_Eav_Block_Adminhtml_Attribute_Edit_Tab_Main extends Mage_Eav_Block_Adminhtml_Attribute_Edit_Main_Abstract
+{
+    /**
+     * Adding product form elements for editing attribute
```
Remove "product"

justinbeaty commented on app/code/core/Mage/Eav/Block/Adminhtml/Attribute/Grid.php:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    /**
+     * Prepare product attributes grid columns
```
Remove "product"

justinbeaty commented on app/code/core/Mage/Eav/Block/Adminhtml/Attribute/Grid/Abstract.php:
```diff
@@ -28,7 +28,7 @@
  * Product attributes grid
```
Remove "product"

justinbeaty commented on app/code/core/Mage/Eav/controllers/Adminhtml/Attribute/AbstractController.php:
```diff
@@ -0,0 +1,346 @@
+
+        $this->_title($id ? $model->getName() : $this->__('New Attribute'));
+
+        $item = $id ? Mage::helper('eav')->__('Edit Product Attribute')
+                    : Mage::helper('eav')->__('New Product Attribute');
```
Remove "product"

justinbeaty commented on app/code/core/Mage/Eav/controllers/Adminhtml/Attribute/AbstractController.php:
```diff
@@ -0,0 +1,346 @@
+
+
+            if ($this->getRequest()->getParam('set') && $this->getRequest()->getParam('group')) {
+                // For creating product attribute on product page we need specify attribute set and group
```
Remove "product"

justinbeaty reviewed (commented):
Some notes to myself

justinbeaty commented:
@fballiano I have use_in_forms working over in #2352. I added a new event to this PR `adminhtml_[entityCode]_attribute_edit_prepare_save` to help accomplish this.

kiatng commented on app/code/core/Mage/Eav/controllers/Adminhtml/Attribute/AbstractController.php:
```diff
@@ -0,0 +1,337 @@
+ * @author      Magento Core Team <core@magentocommerce.com>
+ */
+
+abstract class Mage_Eav_Adminhtml_Attribute_AbstractController extends Mage_Adminhtml_Controller_Action
```
For consistency, is it better to place this in `app/code/core/Mage/Adminhtml/Controller/Eav/Abstract.php`? The code in `Block/Adminhtml` would also need to be placed in `Mage/Adminhtml/Eav`.

justinbeaty replied:
I debated this myself. I recall a discussion somewhere on this GitHub project about how Magento team stuffed everything into Adminhtml because the old code didn't have support for defining adminhtml controllers in the specific module. Magento team later added that support but never broke out the controllers / blocks to the correct module. This was a cleanup project so I didn't want to add to the mess.

justinbeaty replied:
We also already have a directory at `/app/code/core/Mage/Eav/Block/Adminhtml`

kiatng replied:
Thanks for the clarification, I wasn't aware of the history. However I haven't seen abstract class in the `controller` folder. How about placing it in 
`app/code/core/Mage/Eav/Controller/Adminhtml/Attribute/Abstract.php`

justinbeaty replied:
Done, thanks for the suggestion

justinbeaty commented on app/code/core/Mage/Eav/Block/Adminhtml/Helper/Form/Boolean.php:
```diff
@@ -0,0 +1,44 @@
+        parent::__construct($attributes);
+        $this->setValues(array(
+            array(
+                'label' => Mage::helper('catalog')->__('No'),
```
Wrong helper

justinbeaty commented:
I rebased on top of the new 20.0 branch (hopefully all checks pass now), and also moved some new code from #2352 to here

fballiano commented:
@justinbeaty is there still something to be done here? because it would be a real pity to just let all this work die.
I think to be able to edit EAV attributes is fundamental for a good platform.

justinbeaty commented:
@fballiano I was working on the custom form elements so that you can edit pages like customer account info on the frontend without changing templates. It's definitely not dead, I just had less time last few weeks.

I'll pull up the branch today and see where I'm at. I agree it will be a very nice addition and I didn't mean to stall it.

fballiano commented:
now we have some conflicts but @justinbeaty isn't this PR ready to be merged? it's still marked as draft?

because if we abandon this I'll convert my original PRs into modules and release those.

justinbeaty commented:
@fballiano I am so sorry I have had no time to work on this. I do apologize for taking over the work you had started and then not having the time to complete it. Please do take over either the code in this PR or from your original PR. I think it would be good to release it as a module.

I will be back in sometime to continue contributing to OpenMage, just a lot of things have come up lately.

fballiano commented:
I can't take this one over, it's much more complex than my original one, I already invested time into that and I can't do it a second time on this one, which would require much more time.

luigifab commented:
Will be ready for RC4?

fballiano commented:
I wanted to revive this, I think they should target "next" but it's like my preference, because we're more free to choose how to act on the frontend.

fballiano commented:
@justinbeaty do you think you'll have time for this?

fballiano commented:
I'm really not happy about this, I posted two working PRs that got superseeded in a much more complicated way and then abandoned halfway.

This PR is not testable since it's only the base classes.

@kiatng @ADDISON74 @sreichel @Flyingmana we have 2 options IMHO:
- merge this PR with https://github.com/OpenMage/magento-lts/pull/2260 and https://github.com/OpenMage/magento-lts/pull/2264 so that they're testable while fixing them and trying to finish them
- close these ones and restore my previous ones that, as far as I remember, they were easier and working
what do you think?

sreichel commented:
I like the idea and think its worth to complete it. Tried something similiar years ago, but also stop at half way ....

Flyingmana commented:
hmm, I do like to approach this in the more general way as its done in this PR, and in any case I think we should end up there at one point.

But merging it and fixing it on the go requires someone who can commit time to solving it, and in the current situation we do not have anyone being able to provide such guarantee.

So I would go for the previous two working PRs (https://github.com/OpenMage/magento-lts/pull/2260 and https://github.com/OpenMage/magento-lts/pull/2264)

(from here on its not relevant, just a general note) We still can then later move from there towards the goal here, when we have more time available.
With the size of the discussion, I would then also recommend to go smaller steps after the target structure is decided on, and try to cover the relevant parts with Integration tests so we dont need to have UI to test the base parts

ADDISON74 commented:
There are several complex PRs that should have been analyzed separately until the final stage of being ported to OpenMage. Maybe in such cases a repository dedicated to that change is requested.

sreichel commented:
@justinbeaty is it okay for you to merge that other two PRs into - to make it testable?

justinbeaty commented:
> @justinbeaty is it okay for you to merge that other two PRs into - to make it testable?

In hindsight, having three PRs did in fact make this too complicated. It would be easier for me to port https://github.com/MahoCommerce/maho/pull/25 back to OpenMage, but you would need to re-run phpcs, rector, etc.

fballiano commented:
developing this thing is 2 different projects is beyond nonsense.

justinbeaty commented:
> developing this thing is 2 different projects is beyond nonsense.

I don't disagree, it will probably be easier to backport it once it's merged in Maho.

ADDISON74 commented:
I would recommend a separate public discussion. This new e-commerce project named Maho relies on OpenMage platform and not vice versa. The big problem is that Magento 1 didn't offer the 3 PRs out of the box and we had to wait 16 years to discuss them. My opinion is that we should finish them in OpenMage first and then any fork to benefit from it. As far as I am concerned, I choose to contribute to a project where I am not blocked by the owner.

justinbeaty commented:
> I would recommend a separate public discussion. This new e-commerce project named Maho relies on OpenMage platform and not vice versa. The big problem is that Magento 1 didn't offer the 3 PRs out of the box and we had to wait 16 years to discuss them. My opinion is that we should finish them in OpenMage first and then any fork to benefit from it. As far as I am concerned, I choose to contribute to a project where I am not blocked by the owner.

Since the EAV editor was originally @fballiano's idea, I am deciding to contribute the PR to Maho. He is right that it is too much effort to finish developing this feature, bug squashing, etc on two repos. But since this is all Open Source Software, it can be merged here later, or you can continue with the PRs we already have open to OM. I will not be deleting my OM repo that this PR came from.

ADDISON74 commented:
@justinbeaty - The choice is yours. For me it is less important where it is baked, I hope to see functional in OpenMage in the next months.

fballiano commented:
> I would recommend a separate public discussion. 

this is specifically about this PR so why should it be discussed in another place? anyway there's no discussion to make.  so you can delete and lock the discussion as it became usual on this repo.
you have to decide if you want to finish this PR or, as for sure will happen, you will wait for me to finish it, then do the same only to create an incompatibility between the projects.

> This new e-commerce project named Maho relies on OpenMage platform and not vice versa. 

OM has 480 commits made by me (only on the main branch, let's not even calculate the other branches) so don't patronize me on who depends on who. if somebody knows about openmage in 2024 it's only because I spent 3.5 years doing public relations for it, not the other way around.

> My opinion is that we should finish them in OpenMage first and then any fork to benefit from it. 

again, if you want to finish them then do that, but doing the ego race, no thanks, I've 1000 other ideas to develop and opensource is great because, respecting licenses, everybody can benefit.

> As far as I am concerned, I choose to contribute to a project where I am not blocked by the owner.

that's quite obvious, you can't do otherwise anyway

fballiano commented:
> @fballiano ... changing `_getDefaultSourceModel()` to public is a BC break. Just FYI.

do you think that's the only BC i'm making 😂

sreichel commented:
No. But it's a reason to not port the t back.

fballiano commented:
> > it's nice to see that bringing developing something in Maho makes OM wants to merge it immediately after it was left dead for 2 years.
> 
> You are wrong.
> 
> I have re-opened that three PRs a week ago. Two days later @justinbeaty made a PR to your repo.

suuuuuure, it's all a great conincidence.

fballiano commented:
<img width="805" alt="Screenshot 2024-09-24 alle 17 29 31" src="https://github.com/user-attachments/assets/31bb9673-b231-432c-8667-d226f43d0300">

ADDISON74 commented:
Fishing PRs from one side to another is fine as far as I'm concerned, It's already practiced with success. However, I would appreciate if every author would honestly mention where he would like to implement, as happened here. In this way, the code will remain common between repos on certain features. I wish you success in the implementation and I hope to see it merged in OpenMage too.
